### PR TITLE
Bugfix FXIOS-6564 [v116] Error state on credit card number/expiration validation (#14691)

### PR DESCRIPTION
**Pull request by petruSt24**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-6564)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/14691)

## :bulb: Description

changed the data validation to happen on didEndEditing and updated the Save Button state
## :pencil: Checklist
You have to check all boxes before merging
- [X] Filled in the above information (tickets numbers and description of your work)
- [X] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [X] Wrote unit tests and ensured the tests suite is passing
- [ ] Implemented accessibility and tested on UI related work (minimum Dynamic Text and VoiceOver)
- [ ] Updated documentation / comments for complex code and public methods if needed

### CODE REVIEW

File: Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift [hunk 1]
```diff
@@ -169,18 +172,23 @@ struct CreditCardInputField: View {
...
-            .disabled(editMode)
-            .padding(.top, 7.5)
-            .foregroundColor(textFieldColor)
-            .multilineTextAlignment(.leading)
-            .keyboardType(keyboardType)
-            .lineLimit(1)
-            .frame(maxWidth: .infinity, alignment: .leading)
-            .onChange(of: text) { [oldValue = text] newValue in
-                    handleTextInputWith(oldValue, and: newValue)
+        // TODO: to be replaced with FocusState once iOS 14 is dropped
```
Comment (nbhasin2) [inline Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift:175]: nit: Add a ticket related to this todo

**Review by nbhasin2 — CHANGES_REQUESTED** [review]
Screenshot 2023-07-07 at 3 18 01 PM

There are few issues with this approach
- Edit mode doesn't work when you tap on an existing card there top right button to edit is disabled 
- Odd Behaviour: When you add a new card and input incorrect values, then go to next input there is an error but if you try to delete or changes these values then the error still remains and only gets rectified when you go to next input. Not sure if we want that but this could be a UX thing 🤔 

@petruSt24 once you fix this I can take a look at test failures

File: Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift [hunk 2]
```diff
@@ -197,15 +205,49 @@ struct CreditCardInputField: View {
         }
     }
 
+    func handleDidEndEditing(_ textToCheck: String? = nil) {
+        if textToCheck != nil {
+            self.text = textToCheck!
```
Comment (nbhasin2) [inline Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift:210]: @petruSt24 you are checking validity and `textToCheck` is not used anywhere except for name so this seems like a method which requires some finishing touch while `handleTextInputWith` does same or pretty much similar task so we have duplicated code here.

I am wondering if there is something you can do to perform some sort of cleanup?

Comment (nbhasin2) [reply Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift:210]: This is also the reason for your failing tests, could you please cleanup the two methods

Comment (nbhasin2) [reply Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift]: @petruSt24 Pushed a helping commit here for cleanup - https://github.com/mozilla-mobile/firefox-ios/pull/15442/commits/b40006f345d09fec11b24eb801ff3ab03be2cd41

Try to write tests for these as they should be fairly straight forward to test

Comment (petruSt24) [reply Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift]: yes I added that parameter as a helper with the unit tests but it didn't help as the text property can't be changed that way. I did more research about testing swiftUI views and apparently it's not as easy to test as UIKit so I updated the tests to check the viewModel.

**Comment by lmarceau** [comment]
The build failed due to a new warning:

```
⚠️  /Users/vagrant/git/Client/Frontend/Autofill/CreditCard/ViewComponents/CreditCardInputField.swift:180:22: result of call to 'updateInputValidity()' is unused
                self.updateInputValidity()
```

**Comment by nbhasin2** [comment]
Build passed: https://app.bitrise.io/app/6c06d3a40422d10f/pipelines/edbec5f1-3e02-4f86-a0b0-6f18a76fe609

**Pull request merged**